go/vt/vttablet/tabletserver/state_manager_test.go

```diff
@@ -0,0 +1,604 @@
+	assert.Equal(t, int64(StateServing), sm.State())
+}
+
+func TestStateManagerValidations(t *testing.T) {
+	sm := newTestStateManager(t)
+	target := &querypb.Target{TabletType: topodatapb.TabletType_MASTER}
+	sm.target = *target
+
+	err := sm.StartRequest(ctx, target, false)
+	assert.Contains(t, err.Error(), "operation not allowed")
+
+	sm.state = StateServing
+	sm.transitioning = true
+	err = sm.StartRequest(ctx, target, false)
+	assert.Contains(t, err.Error(), "operation not allowed")
+
+	err = sm.StartRequest(ctx, target, true)
+	assert.NoError(t, err)
+
+	sm.transitioning = false
+	target.Keyspace = "a"
+	err = sm.StartRequest(ctx, target, false)
+	assert.Contains(t, err.Error(), "invalid keyspace")
+	err = sm.VerifyTarget(ctx, target)
+	assert.Contains(t, err.Error(), "invalid keyspace")
+
+	target.Keyspace = ""
+	target.Shard = "a"
+	err = sm.StartRequest(ctx, target, false)
+	assert.Contains(t, err.Error(), "invalid shard")
+	err = sm.VerifyTarget(ctx, target)
+	assert.Contains(t, err.Error(), "invalid shard")
+
+	target.Shard = ""
+	target.TabletType = topodatapb.TabletType_REPLICA
+	err = sm.StartRequest(ctx, target, false)
+	assert.Contains(t, err.Error(), "invalid tablet type")
+	err = sm.VerifyTarget(ctx, target)
+	assert.Contains(t, err.Error(), "invalid tablet type")
+
+	sm.alsoAllow = []topodatapb.TabletType{topodatapb.TabletType_REPLICA}
+	err = sm.StartRequest(ctx, target, false)
+	assert.NoError(t, err)
+	err = sm.VerifyTarget(ctx, target)
+	assert.NoError(t, err)
+
+	err = sm.StartRequest(ctx, nil, false)
+	assert.Contains(t, err.Error(), "No target")
+	err = sm.VerifyTarget(ctx, nil)
+	assert.Contains(t, err.Error(), "No target")
+
+	localctx := tabletenv.LocalContext()
+	err = sm.StartRequest(localctx, nil, false)
+	assert.NoError(t, err)
+	err = sm.VerifyTarget(localctx, nil)
+	assert.NoError(t, err)
+}
+
+func TestStateManagerWaitForRequests(t *testing.T) {
+	sm := newTestStateManager(t)
+	target := &querypb.Target{TabletType: topodatapb.TabletType_MASTER}
+	sm.target = *target
+	sm.timebombDuration = 10 * time.Second
+
+	_, err := sm.SetServingType(topodatapb.TabletType_MASTER, StateServing, nil)
+	require.NoError(t, err)
+
+	err = sm.StartRequest(ctx, target, false)
+	require.NoError(t, err)
+
+	// This will go into transition and wait.
+	// Wait for that state.
+	go sm.StopService()
+	for {
+		sm.mu.Lock()
+		transitioning := sm.transitioning
+		sm.mu.Unlock()
+		if !transitioning {
+			continue
+		}
+		time.Sleep(10 * time.Millisecond)
+		break
+	}
+
+	// Verify that we're still transitioning.
+	sm.mu.Lock()
+	assert.True(t, sm.transitioning)
+	sm.mu.Unlock()
+
+	sm.EndRequest()
+
+	for {
+		sm.mu.Lock()
+		transitioning := sm.transitioning
+		sm.mu.Unlock()
+		if transitioning {
+			time.Sleep(10 * time.Millisecond)
+			continue
+		}
+		break
+	}
+	assert.Equal(t, int64(StateNotConnected), sm.State())
+}
+
+func verifySubcomponent(t *testing.T, component interface{}, order int64, state testState) {
```

Comment:
super-nit: I think it would be nice if we changed the order of these, so that `order` comes before `component`. That way, assertions line up much better:
```
verifySubcomponent(t, 1, sm.messager,  testStateClosed)
verifySubcomponent(t, 2, sm.te,  testStateClosed)
verifySubcomponent(t, 3, sm.txThrottler,  testStateClosed)
verifySubcomponent(t, 4, sm.qe,  testStateClosed)
verifySubcomponent(t, 5, sm.watcher,  testStateClosed)
...
```